content/doc/book/pipeline/scaling-pipeline.adoc

```diff
@@ -86,6 +86,14 @@ A dirty shutdown can also happen due to catastrophic operating system failures,
 
 == Other Scaling Suggestions
 
+* Use @NonCPS-annotated functions for more complex work. This means more involved processing, logic, and transformations. This lets you leverage additional Groovy & functional features for more powerful, concise, and performant code.
+** This still runs on masters so be aware of complexity of the work, but is much faster than native Pipeline code because it doesn’t provide durability and uses a faster execution model. Still, be mindful of the CPU cost and offload to executors when the cost becomes too high.
+** @NonCPS functions can use a much broader subset of the Groovy language, such as iterators and functional features, which makes them more terse and fast to write.
```

Comment:
I'd have to check which features specifically still don't work, since the DefaultGroovyMethods handles most of 'em now -- maybe just say that if Groovy features don't work, it's worth trying NonCPS here?

Reply:
@svanoort Ugh.  Please no.  If we suggest people "try" something.  They're going to use in all kinds of places it isn't needed.  (I did.)

Reply:
@bitwiseman That's actually a good idea in this case as long as you're only applying it to pure-Groovy functions (no step calls).  It doesn't have any harmful effects and improves performance and memory use and makes the code less prone to odd results. 

But yeah I get your point that it's not a good general purpose in normal cases.